packages/models/src/store/molecule-storage.ts

```diff
@@ -0,0 +1,303 @@
+  | VecCodecConfig
+  | TableCodecConfig
+  | OptionCodecConfig
+  | UnionCodecConfig
+
+type GetFixedCodecConfig<T extends FixedParam, Deep extends DeepLength[number] = 6> = Deep extends never
+  ? never
+  : T extends FixedBasic
+  ? GetOneType<T>
+  : T extends ArrayParam
+  ? {
+      type: GetMoleculeType<'array'>
+      value: IsConstNumber<T['value'][1]> extends never
+        ? never
+        : GetFixedCodecConfig<T['value'][0], GetPreDeep<Deep>> extends never
+        ? never
+        : Tuple<GetFixedCodecConfig<T['value'][0], GetPreDeep<Deep>>, T['value'][1]>
+    }
+  : T extends StructParam
+  ? {
+      type: GetMoleculeType<'struct'>
+      value: { [P in keyof T['value']]: GetFixedCodecConfig<T['value'][P], GetPreDeep<Deep>> }
+    }
+  : never
+
+export type GetCodecConfig<T extends DynamicParam, Deep extends DeepLength[number] = 6> = Deep extends never
+  ? never
+  : T extends FixedBasic | DynamicBasic
+  ? GetOneType<T>
+  : T extends FixedParam
+  ? GetFixedCodecConfig<T, GetPreDeep<Deep>>
+  : T extends VecParam
+  ? { type: GetMoleculeType<'vec'>; value: GetCodecConfig<T['value'], GetPreDeep<Deep>> }
+  : T extends TableParam
+  ? {
+      type: GetMoleculeType<'table'>
+      value: { [P in keyof T['value']]: GetCodecConfig<T['value'][P], GetPreDeep<Deep>> }
+    }
+  : T extends OptionParam
+  ? { type: GetMoleculeType<'option'>; value: GetCodecConfig<T['value'], GetPreDeep<Deep>> }
+  : T extends UnionParam
+  ? {
+      type: GetMoleculeType<'union'>
+      value: { [P in keyof T['value']]: GetCodecConfig<T['value'][P], GetPreDeep<Deep>> }
+    }
+  : never
+
+type GetFixedOffChain<T extends FixedParam, Deep extends DeepLength[number] = 6> = Deep extends never
+  ? never
+  : T extends FixedBasic
+  ? GetBasicOffChain<T>
+  : T extends StructParam
+  ? { [P in keyof T['value']]: GetFixedOffChain<T['value'][P], GetPreDeep<Deep>> }
+  : T extends ArrayParam
+  ? Tuple<GetFixedOffChain<T['value'][0], GetPreDeep<Deep>>, T['value'][1]>
+  : never
+
+export type GetMoleculeOffChain<T extends DynamicParam, Deep extends DeepLength[number] = 6> = Deep extends never
+  ? never
+  : T extends FixedBasic | DynamicBasic
+  ? GetBasicOffChain<T>
+  : T extends FixedParam
+  ? GetFixedOffChain<T, GetPreDeep<Deep>>
+  : T extends VecParam
+  ? GetMoleculeOffChain<T['value'], GetPreDeep<Deep>>[]
+  : T extends TableParam
+  ? { [P in keyof T['value']]: GetMoleculeOffChain<T['value'][P], GetPreDeep<Deep>> }
+  : T extends OptionParam
+  ? GetMoleculeOffChain<T['value'], GetPreDeep<Deep>>
+  : T extends UnionParam
+  ? OneOfRecord<{ [P in keyof T['value']]: GetMoleculeOffChain<T['value'][P], GetPreDeep<Deep>> }>
+  : never
+
+export function isFixedCodecConfig(value: unknown): value is GetCodecConfig<FixedParam> {
+  if (typeof value === 'string') return ([...FixedBasicTypes] as string[]).includes(value)
+  if (
+    typeof value === 'object' &&
+    value !== null &&
+    'type' in value &&
+    typeof value.type === 'string' &&
```

Comment:
`typeof value?.type === 'string'` would be enough

Reply:
Because the value is `unknown`, we should narrow its type by
```
typeof value === 'object' &&
    value !== null &&
    'type' in value 
```